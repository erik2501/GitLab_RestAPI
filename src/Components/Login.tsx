import { useEffect, useState } from 'react';
import { Card, CardContent, CardHeader, CardActions, TextField, Button } from '@mui/material';
import { getProjectInfo } from '../helpers/fetches';
import { Project } from "../helpers/types";
import { useLogin } from './ProjectContext';


const Login = () => {

    const projectContext = useLogin();
    const [projectID, setProjectID] = useState<string>();
    const [token, setToken] = useState<string>();
    const [message, setMessage] = useState<string>('');
    const [projectData, setProjectData] = useState<Project>();
    const [teamName, setTeamName] = useState<string | null>()

    useEffect(() => {
        if (projectData?.id !== -1 && projectID && token) {
            projectContext?.setProject({
                projectID: projectID,
                token: token
            })
            if (projectData) {
                setTeamName(projectData.namespace.name);
                localStorage.setItem('teamName', projectData.namespace.name);
            }
        } else {
            setMessage('Wrong login info, try again.');
        }
    }, [projectData])

    useEffect(() => {
        const lsProjectID = localStorage.getItem('projectID');
        const lsToken = localStorage.getItem('token');
        setTeamName(localStorage.getItem('teamName'));
        if (lsProjectID && lsToken) {
            projectContext?.setProject({
                projectID: lsProjectID,
                token: lsToken
            })
            setProjectID(lsProjectID);
            setToken(lsToken);
        }
        setMessage('');
    }, [])

    const logIn = () => {
        if (projectID && token) {
            getProjectInfo(projectID, token).then(data => setProjectData(data));
        } else {
            setMessage('You have to fill in all forms');
        }
    }

    const logOut = () => {
        projectContext?.setProject(null);
        setMessage('');
        setProjectID('');
        setToken('');
        localStorage.setItem('projectID', '');
        localStorage.setItem('token', '');
    }


    if (projectContext?.project) {
        return (
            <div className='parentcontainer'>
                <h1>Welcome to {teamName}!</h1>
                <Button
                    onClick={logOut}
                    className='loginBtn'
                    variant='contained'
                    style={{ backgroundColor: '#55828b' }}>
                    Log out
                </Button>
            </div>
        )
    }
    else {
        return (
            <div className='parentcontainer'>
                <form className='logincontainer'>
                    <Card className='loginCard' sx={{ width: '100%', boxShadow: '10' }}>
                        <div className='parentcontainer'>
                            <CardHeader className='cardHeaderLogin' title="Find your Git-repository"  />
                        </div>
                        <CardContent className='loginTextFields' >
                            <div>
                                <TextField
                                    id="projectID"
                                    label="Project ID"
                                    placeholder='Link'
                                    value={projectID}
                                    onChange={(event) => { setProjectID(event.target.value); }}
                                />
                            </div>
                            <div>
                                <TextField
                                    id="access-token"
                                    label="Access Token"
                                    placeholder='Token'
                                    value={token}
                                    onChange={(event) => { setToken(event.target.value); }}
                                />
                            </div>
                        </CardContent>
                        <CardActions style={{ display: 'flex', justifyContent: 'center' }}>
                            <div>
                                <Button
                                    data-testid='loginBtn'
                                    className='loginBtn'
                                    // We have used inline styling here because this is a mui button
                                    style={{ maxWidth: '100%', display: 'flex', justifyContent: 'center', backgroundColor: '#55828b' }}
                                    variant='contained'
                                    onClick={logIn}>
                                    Find Git-repository
                                </Button>
                                <p>{message}</p>
                            </div>
                        </CardActions>
                    </Card>
                </form>
            </div>
        )
    }
}


export default Login;