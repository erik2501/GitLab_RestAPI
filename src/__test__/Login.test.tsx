import Login from "../Components/Login";
import renderer from 'react-test-renderer'
import { fireEvent, render, screen } from "@testing-library/react";


it('when run the login renders', () => {
    const tree = renderer.create((<Login  />)).toJSON();
    expect(tree).toMatchSnapshot();
});

test('check login button when empty fields', () => {
    render(<Login></Login>)
    fireEvent.click(screen.getByTestId('loginBtn'))
    screen.getByText('You have to fill in all forms')
});