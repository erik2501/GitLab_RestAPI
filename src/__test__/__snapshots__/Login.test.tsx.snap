// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when run the login renders 1`] = `
<div
  className="parentcontainer"
  style={
    Object {
      "paddingTop": "40px",
    }
  }
>
  <form
    className="logincontainer"
  >
    <div
      className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-tjaq38-MuiPaper-root-MuiCard-root"
    >
      <div
        className="parentcontainer"
      >
        <div
          className="MuiCardHeader-root css-185gdzj-MuiCardHeader-root"
          style={
            Object {
              "display": "flex",
              "justifyText": "center",
            }
          }
        >
          <div
            className="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
          >
            <span
              className="MuiTypography-root MuiTypography-h5 MuiCardHeader-title css-1qvr50w-MuiTypography-root"
            >
              Find your Git-repository
            </span>
          </div>
        </div>
      </div>
      <div
        className="MuiCardContent-root css-46bh2p-MuiCardContent-root"
        style={
          Object {
            "columnGap": "10px",
            "display": "flex",
            "justifyContent": "center",
          }
        }
      >
        <div>
          <div
            className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
          >
            <label
              className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink={false}
              htmlFor="projectID"
              id="projectID-label"
            >
              Project ID
            </label>
            <div
              className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
              onClick={[Function]}
            >
              <input
                aria-invalid={false}
                autoFocus={false}
                className="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                disabled={false}
                id="projectID"
                onAnimationStart={[Function]}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="Link"
                required={false}
                type="text"
              />
              <fieldset
                aria-hidden={true}
                className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  className="css-yjsfm1"
                >
                  <span>
                    Project ID
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div>
          <div
            className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
          >
            <label
              className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink={false}
              htmlFor="access-token"
              id="access-token-label"
            >
              Access Token
            </label>
            <div
              className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
              onClick={[Function]}
            >
              <input
                aria-invalid={false}
                autoFocus={false}
                className="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                disabled={false}
                id="access-token"
                onAnimationStart={[Function]}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="Token"
                required={false}
                type="text"
              />
              <fieldset
                aria-hidden={true}
                className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  className="css-yjsfm1"
                >
                  <span>
                    Access Token
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
      </div>
      <div
        className="MuiCardActions-root MuiCardActions-spacing css-1t6e9jv-MuiCardActions-root"
        style={
          Object {
            "display": "flex",
            "justifyContent": "center",
          }
        }
      >
        <div>
          <button
            className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium loginbtn css-sghohy-MuiButtonBase-root-MuiButton-root"
            data-testid="loginBtn"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            style={
              Object {
                "display": "flex",
                "justifyContent": "center",
                "maxWidth": "100%",
              }
            }
            tabIndex={0}
            type="button"
          >
            Find Git-repository
          </button>
          <p />
        </div>
      </div>
    </div>
  </form>
</div>
`;
